pkg/storage/unified/indexing/main.go

```diff
@@ -0,0 +1,340 @@
+	}
+
+	// Display search results
+	fmt.Printf("Total Hits: %d\n", searchResult.Total)
+	for _, hit := range searchResult.Hits {
+		fmt.Printf("Object ID: %s, Score: %f\n", hit.ID, hit.Score)
+
+		// Retrieve the document to display more details
+		//var doc Object
+		//err := index.Object(hit.ID).VisitFields(func(field bleve.DocumentField) {
+		//	switch field.Name() {
+		//	case "Title":
+		//		doc.Title = string(field.Value())
+		//	case "Content":
+		//		doc.Content = string(field.Value())
+		//	case "Tags":
+		//		doc.Tags = append(doc.Tags, string(field.Value()))
+		//	}
+		//})
+		//if err != nil {
+		//	log.Printf("Failed to retrieve document %s: %v", hit.ID, err)
+		//	continue
+		//}
+		//
+		//fmt.Printf("Title: %s\n", doc.Title)
+		//fmt.Printf("Content: %s\n", doc.Content)
+		//fmt.Printf("Tags: %v\n", doc.Tags)
+		//fmt.Println("---------")
+	}
+}
+
+func seedObjects() []Object {
+	type Resource struct {
+		Group      string
+		Resource   string
+		Kind       string
+		ApiVersion string
+	}
+
+	namespaces := []string{"stack-1", "stack-2"}
+	resources := []Resource{
+		{
+			Group:      "playlist.grafana.app",
+			Resource:   "playlists",
+			Kind:       "Playlist",
+			ApiVersion: "playlist.grafana.app/v0alpha1",
+		},
+		{
+			Group:      "folder.grafana.app",
+			Resource:   "folders",
+			Kind:       "Folder",
+			ApiVersion: "folder.grafana.app/v0alpha1",
+		},
+	}
+
+	// creates 4,000,000 objects
+	objects := []Object{}
+	for _, namespace := range namespaces {
+		for _, resource := range resources {
+			for i := 0; i < 10000; i++ {
+				objects = append(objects, Object{
+					Guid:            fmt.Sprintf("%s-%s-%d", namespace, resource.Resource, i),
+					ResourceVersion: time.Now().UnixNano(),
+					Group:           resource.Group,
+					Resource:        resource.Resource,
+					Namespace:       namespace,
+					Name:            fmt.Sprintf("%s-%s-%d", namespace, resource.Resource, i),
+					Value: K8sObject{
+						Kind:       resource.Kind,
+						ApiVersion: resource.ApiVersion,
+						K8sMeta: K8sMeta{
+							Name:              fmt.Sprintf("%s-%s-%d", namespace, resource.Resource, i),
+							Namespace:         namespace,
+							Uid:               fmt.Sprintf("k8s-uid-%d", i),
+							CreationTimestamp: time.Now().Add(-time.Duration(i) * time.Hour).Format(time.RFC3339),
+						},
+					},
+				})
+			}
+		}
+	}
+
+	return objects
+}
+
+func IndexObjects(index bleve.Index, objects []Object) {
+	for _, obj := range objects {
+		err := index.Index(obj.Guid, obj)
+		if err != nil {
+			fmt.Println("Failed to index document:", err)
+		}
+	}
+
+	fmt.Println("Indexing synchronously completed")
+}
+
+func indexObjectsConcurrently(index bleve.Index, objects []Object) {
+	// Create a wait group to wait for all goroutines to finish
+	var wg sync.WaitGroup
+
+	// Create a channel to send batches to the workers
+	batchChan := make(chan *bleve.Batch)
+
+	// Start workers (goroutines)
+	for i := 0; i < INDEX_WORKERS; i++ {
```

Comment:
i don't think it's necessary to loop over and create 4000 workers.  probably won't take too long but there is some overhead.  you could just call a go func in your loop below and remove this.  something like

```
func main() {
    data := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
    chunkSize := 3

    chunks := chunkSlice(data, chunkSize)

    var wg sync.WaitGroup
    for _, chunk := range chunks {
        wg.Add(1)
        go func(c []int) {
            defer wg.Done()
            // Process the chunk concurrently
            fmt.Println(c)
        }(chunk)
    }
    wg.Wait()
}

func chunkSlice(slice []int, chunkSize int) [][]int {
    var chunks [][]int
    for i := 0; i < len(slice); i += chunkSize {
        end := i + chunkSize
        if end > len(slice) {
            end = len(slice)
        }
        chunks = append(chunks, slice[i:end])
    }
    return chunks
}
```

Reply:
oh i missed that you were creating a pool of workers with INDEX_WORKERS.  i guess that does give you better control over how many goroutines are created